### PR TITLE
[Nanoleaf] Visualize layout

### PR DESCRIPTION
**Pull request by austvik**
When adding a Nanoleaf device to your Openhab, each light panel will get a separate thing, identified by its ID. It is hard to figure out which ID is which panel, especially if you have many panels. The Nanoleaf binding has had a way to print the layout with the ids for the original type of Nanoleaf panels, but no good way to identify what is what for more modern panel types.

This pull request adds a Layout channel on the controller which returns an image with the panel ids, so that it is easier to find the IDs you are interested in.

This is only tested on Nanoleaf Elements Hexagons, but should work on many other shapes and types of panels.

Signed-off-by: Jørgen Austvik [jaustvik@acm.org](mailto:jaustvik@acm.org)

### CODE REVIEW

**Comment by jlaur** [comment]
@austvik - this is a very creative approach to solving that problem. :-) I would be curious to see how it looks.

However, I'm also thinking if creating such a channel for temporary use to identify IDs is the right approach. Perhaps you could have a look at #13555 which seems to solve a quite similar problem in a different way. Don't read this as I have anything against your approach, just consider it for inspiration. If I'm not mistaken and the problem is indeed similar, perhaps it would be worth settling on some common approach. In this case I would personally consider the console approach less bloated and more streamlined. WDYT?

Cc @lolodomo

**Comment by lsiepel** [comment]
> @austvik - this is a very creative approach to solving that problem. :-) I would be curious to see how it looks.
> 
> However, I'm also thinking if creating such a channel for temporary use to identify IDs is the right approach. Perhaps you could have a look at #13555 which seems to solve a quite similar problem in a different way. Don't read this as I have anything against your approach, just consider it for inspiration. If I'm not mistaken and the problem is indeed similar, perhaps it would be worth settling on some common approach. In this case I would personally consider the console approach less bloated and more streamlined. WDYT?
> 
> Cc @lolodomo

If there is a need for a common approach isn't this visual solution based on a channel easier to use than a more techy console solution?

**Comment by jlaur** [comment]
@austvik - could you provide a screenshot from this?

@lsiepel - I agree this approach is more visual and perhaps even more user-friendly. Maybe I was being a bit too harsh with my comment. :-) IMHO we can merge the PR as is. However, moving forward, it might make sense to harmonize the way of finding IDs which can otherwise be hard to find.

Please correct me here if I'm wrong:
- User creates an item linked to the new channel.
- IDs can now be seen from MainUI.
- The IDs of interest are written down (copy'n'paste not possible).
- The item is removed again to clean up.

With console:
- Console is started.
- Command is executed.
- The IDs of interest can be copied.
- Console is exited.

With the console approach you can even use SSH from remote and there is no runtime penalty. I haven't seen the visual approach yet, but I could imagine if more bindings would follow this road, there would be quite different visualizations with different layout, colors etc. Not saying it's a problem, just stating it probably won't be as neutral or "to the point" as a console command.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/handler/NanoleafControllerHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/handler/NanoleafControllerHandler.java:713]: I would suggest to return fast from this method if no items are linked to the channel. This would reduce the runtime cost.

**Comment by austvik** [comment]
image
This is how my setup looks

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/handler/NanoleafControllerHandler.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/handler/NanoleafControllerHandler.java:713]: I was thinking about this tonight, and updateState() all over OpenHab should maybe oinly take a java.util.function.Function which is only called if a channel was linked to something.

**Comment by austvik** [comment]
The layout implementation returns a byte[], so can easily be written to a file from the console. I just dont think it is user friendly. ssh in, find the users, find some directory the software has write access to, scp it back.

**Comment by lsiepel** [comment]
Would it be possible to have this as a bridge property instead of a channel?

**Comment by austvik** [comment]
No. Properties are strings in OpenHab.

    public Map<String, String> getProperties();
    public String setProperty(String string, String string1);
    public void setProperties(Map<String, String> map);

**Comment by lsiepel** [comment]
IMHO would be the best place if you ask me. Not sure if it would ever be possible to also support images.

I see more usecases like solarpanel layout etc.

**Comment by austvik** [comment]
I totally agree. If properties were <String, **State**> instead of <String, String>, that is where it would fit.

**Comment by austvik** [comment]
image

Added some style to the image and only calculate the image if the channel is linked.

**Comment by stefan-hoehn** [comment]
First of all, @austvik, this looks **awesome**. This is what I always wanted to do and I even started a branch on that many months ago but due to the blockly work I always pushed that back. So great YOU did it and much nicer than I did!

As you you probably know there IS already a console command available: see https://www.openhab.org/addons/bindings/nanoleaf/#panel-layout
but that only does it with characters and also it only supports squares.

I haven't yet understood how your channel idea works (I have not yet had time to look into the code) but my first implementation of that went by a channel as well and it was actually "rejected" by @kaikreuzer by then as channels are not supposed to be used for that. So give me a few days to look into it and to even try it out and will come back to you.

Also let me note that I got a documentation from nanoleaf how to "officially" calculate all shapes (it is actually a few handwritten pages full of math :-) ). I am happy to forward that to you.

So again, great that you did it. Just give me a few days (I am off next week and away from home with nanoleads), so expect something in around 2 weeks, okay? Btw, I have four different sets of panels and shapes to test everything here.

CU soon
Stefan

**Comment by austvik** [comment]
image
image

A couple of screenshots to show how the channel work.

I'm interrested in your document (jaustvik@acm.org), and would also love to get a copy of your JSON (panelLayout with globalOrientation and layout) to test the algorithm.

**Comment by stefan-hoehn** [comment]
Expect a mail in your inbox just now (you may want to edit you comment above and remove the mail address if you like)

**Comment by kaikreuzer** [comment]
> my first implementation of that went by a channel as well and it was actually "rejected" by @kaikreuzer by then

This sounds too hard. 😉 We had a similar discussion with arguments @jlaur mentions above - usually the console is quite helpful for setting up things, so that was my advice at that time. Seeing @austvik's nice images and the fact that the hexagons are fairly difficult to render in ascii art, I'd also think that the channel on the bridge can be a suitable solution here (while indeed it is a bit of a misuse of channels). But creating an image on the file-system through the console is indeed no option as it would be hard to transfer it to a place where the user can actually look at it.

> The item is removed again to clean up.

@jlaur I'd expect in practise that users will simply leave the item in place. They can at any time look at the layout and also have it updated whenever it changes. The implementation should merely make sure to not send item updates too often as this will mean quite some traffic on the event bus...

So from my pov: Let's go ahead with that solution, I like it! 👍

**Comment by jlaur** [comment]
> > The item is removed again to clean up.
> 
> @jlaur I'd expect in practise that users will simply leave the item in place. They can at any time look at the layout and also have it updated whenever it changes. The implementation should merely make sure to not send item updates too often as this will mean quite some traffic on the event bus...

In the current implementation it will be updated each minute (with the default configured refresh interval) if the channel is linked to any items (i.e. when not cleaning up after use). @austvik - it might be worth considering if this could be triggered on actual changes. I have not checked if this is feasible, but I'll await your reply before merging.

> So from my pov: Let's go ahead with that solution, I like it! 👍

Great, so we have consensus to accept and merge this. :-) After seeing the actual rendered result, it seems more complex that I had imagined, so I agree that the graphical visualization certainly adds value over the console approach for this case.

**Comment by austvik** [comment]
I've added changes so that it only renders the layout if the layout has changes (and is linked to an item), and it now supports triangles, hexagons and squares. 

image
image
image
image

**Comment by stefan-hoehn** [comment]
Thanks for checking my layouts as well. I love that. I haven't yet looked at the code but in general I love what it creates! I also love the fact that it only happens when the layout changes and that you have added the panel id's to it. Btw, it seems my panels don't seem to be hanging completely right on the wall (looking on the non-straight lines) ;-)

Can you please add some documentation in the Readme.me how the channel works, how it is supposed to be used, how it could be used in the UI and at best with an example (even with a picture) because this is how and why people start using it.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java
Comment (stefan-hoehn) [inline bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java:91]: Would it be possible to use draw with a polygon shape, so in the future we could even fill that shape with a particular color? (just asking ...)

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java
Comment (stefan-hoehn) [inline bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java:110]: I only now discover that you use the .CORNER type as well (you'll see other comments below because I only review this part as the last piece of code after having gone through all the others). However, I don't understand what the corner type is for. Can you elaborate on that or even document it in the code?

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java
Comment (stefan-hoehn) [inline bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java:37]: The master piece of all 💯 

Btw, there is a type N - O - noleaf ...

**Review by stefan-hoehn — CHANGES_REQUESTED** [review]
First of all, don't be frightened or even annoyed about the many comments, rather take it as an appreciation of all the great effort you have done to the Nanoleaf binding. This is something I always wanted to do and you finally delivered it and probably much better than I would have done!
Thanks for all your hard work,
Stefan

File: bundles/org.openhab.binding.nanoleaf/README.md
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/README.md:97]: Thanks! I have tried to add some more documentation and a picture of my layout. I didn't add anything from the OpenHAB UI so that the picture stays the same even though the UI may change.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/NanoleafBindingConstants.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/NanoleafBindingConstants.java:82]: I have added the ones I could find SKUs for on the README and 47/48 to touch support

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java:91]: Ive been thinking a little about color. I think for that we should add another channel which has color but not the thing IDs(?) For the shapes it is pretty easy, but for the elements where each corner has an individual light it is more interesting because you would have to blend colors in triangles - one for each side of the hexagon - so I dont think we will draw polygons then either, or maybe 6 of them with separate Paints.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java:110]: Good idea! Added a comment. They are very different from the traditional shapes.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java:37]: Actually, I have been able to move most of it out to the layout.shape package, except the corners. Maybe a more in line design for the corners bubbles out if we add something for rendering colors. Maybe the factory should take in a stream of positionDatum, and based on the shape type it decides how many it needs to read from it or something.

**Comment by austvik** [comment]
Thanks for the review comments! Will fix. GitHub hid a lot of them, so I need another pass.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/ShapeType.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/ShapeType.java:30]: Thanks, done!

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/shape/ShapeFactory.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/shape/ShapeFactory.java:40]: I Dont have any data for them, but if I get some JSON, Ill try.

Corners are implemented (the first screenshot are of elements corners, but they are different from the other shapes. Added a comment about it in the layout code.)

File: bundles/org.openhab.binding.nanoleaf/README.md
Comment (stefan-hoehn) [inline bundles/org.openhab.binding.nanoleaf/README.md:79]: typos:  
induvidual -> individual
thing IDs -> thing ID

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/Point2D.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/Point2D.java:19]: He, yeah it grew :) Changed it a bit.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/shape/ShapeFactory.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/shape/ShapeFactory.java:40]: Corners are implemented in NanoleafLayout. Might rework how they are done a little if we want pictures with colors. I think it is possible to make them more aligned to the other shapes. Need some JSON to draw lines.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java:172]: Sorry, I dont understand this comment.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java:175]: Sorry, I dont understand this comment.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/DrawingAlgorithm.java
Comment (stefan-hoehn) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/DrawingAlgorithm.java:28]: Maybe you want to ask in the forum if someone can contribute them but of course not at all blocking for the merge ;-)

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java:170]: Ah, then I maybe understand the other comments. It is used to detect if there is a change. Lets say you have the old layout data and the new layout data. If they are both null, then there is nothing to draw. It was nothing to draw before, and is nothing to draw now. If the old one was null, and the new one is not null, we want to draw the image. And if none of them are null, we need to check down the object structure.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/PositionDatum.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/PositionDatum.java:101]: I needed more data and didn't see this one before after I added it.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/layout/NanoleafLayout.java]: Added

File: bundles/org.openhab.binding.nanoleaf/src/main/resources/OH-INF/i18n/nanoleaf.properties
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/resources/OH-INF/i18n/nanoleaf.properties:42]: It said panels elsewhere, so changed it to thar

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java:178]: The philosophical question is "is null equal to null".
The practical question is "should we render the picture if the previous layout was null and the new layout is null".

I could change equal to be false if any was null, and have null checking in the layout instead (it is probably already there).

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java
Comment (stefan-hoehn) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java:170]: Yes the comments >//< where meant as inline Java comments that should show what I is happening at that point, so I can understand the intent or what is meant to happen. 
Maybe it is only me but I feel the code (even if it may be correct) is hard to understand in terms what its intention is. I tried to tell what I think it is doing which leads me to telling you,  I don't get it. 

Maybe some comments would help and adding a few tests in the test class that show what is meant to happen.

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java
Comment (stefan-hoehn) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java:172]: see comment above (it is mean as a description what I think is happening here)

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java
Comment (stefan-hoehn) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java:175]: see comment above (it is mean as a description what I think is happening here)

File: bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java
Comment (stefan-hoehn) [reply bundles/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/model/Layout.java:178]: I recommend commenting the code a bit and adding a test case that clarifies what you intend to do.

**Comment by stefan-hoehn** [comment]
@austvik It seems you are done with the changes, right? So far everything looks good to me I would then (try to) approve the changes.

**Comment by austvik** [comment]
The thing that is not fixed is the null checking in the equals. I think it is correct, but it is written a little but strange to make the compiler understand that some @Nullable variable can't be null. If you think it is OK, then we can go with it, if not I can put the "both null => return" case first?

**Comment by stefan-hoehn** [comment]
I actually trust you there as I've reviewed the full code you provided pretty thoroughly which to me looks overall really good. Whatever way you feel is good, let's go for that.

**Comment by austvik** [comment]
OK, then I think it would be good to get this in. Ill see if I can do something for the colors this weekend. I might be able to generalize the corners then. I think getting colors for shapes is much easier than for elements corners.

File: bundles/org.openhab.binding.nanoleaf/README.md
Comment (austvik) [reply bundles/org.openhab.binding.nanoleaf/README.md:73]: I think we can leave that to an independant commit?

**Comment by austvik** [comment]
I have tried to rewrite the hard equals methods. From my POV I think this is ready now? - but I am not familiar with the process.

File: bundles/org.openhab.binding.nanoleaf/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.nanoleaf/README.md:77]:
```suggestion
If you want to program individual panels, it can be hard to figure out which panel has which ID. To make this easier, there is Layout channel on the Nanoleaf controller thing in openHAB. 
```

File: bundles/org.openhab.binding.nanoleaf/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.nanoleaf/README.md:78]:
```suggestion
The easiest way to visualize the layout of the individual panels is to open the controller thing in the openHAB UI, go to Channels and add a new item to the Layout channel. 
```

File: bundles/org.openhab.binding.nanoleaf/src/main/resources/OH-INF/i18n/nanoleaf_de.properties
Comment (jlaur) [inline bundles/org.openhab.binding.nanoleaf/src/main/resources/OH-INF/i18n/nanoleaf_de.properties:2]: Please remove the changes in German translation. Changes must be provided through Crowdin after the PR has been merged. See https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations.

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for your contribution! I have added only some very minor comments.

**Comment by austvik** [comment]
Thank you, Jacob. I believe your comments are now fixed.

**Comment by jlaur** [comment]
@austvik - something went wrong with last commit: A new file `HexagonCorners` was added, containing duplicate class `Hexagon`. Also, the changes in `nanoleaf_de.properties` must be completely reverted.

**Comment by austvik** [comment]
Thanks @jlaur! I've removed the extra file and the unintentional changes to the German translation.

**Review by jlaur — APPROVED** [review]
LGTM - thanks!

**Comment by jlaur** [comment]
@austvik - thanks again. Perhaps PR title could be rephrased slightly, for example: "Add channel for visualizing layout"? That would give users reading release notes a direct hint how to utilize this without further reading.

**Comment by austvik** [comment]
Thanks everybody!

And good suggestion on the PR title - I have updated it.

**Comment by stefan-hoehn** [comment]
A final "thanks" for the contribution - I love it.

**Comment by stefan-hoehn** [comment]
FYI: it has become part of https://github.com/openhab/openhab-distro/releases/tag/3.4.0.M5

**Pull request merged**
